M2ASDA offers a variety of functionalities for single-cell omics data analysis, including but not limited to: 1) detecting anomalous cells 2) aligning multi-sample batch cells 3) subtyping anomalous cells across multi-batch. Additionally, all the functions can be implemented through both python package and terminal commands.

## Preparations before tutorials
Before starting the tutorial, we need to make some preparations, including: installing M2ASDA and its required Python packages, downloading the datasets required for the tutorial, and so on. The preparations is available at [M2ASDA Preparations](../start.md). Additionally, when dealing with multimodal data structures involving large gene expression matrices, we strongly recommend using a GPU to train M2ASDA for faster execution.


## Outline of tutorials
- [Tutorial 1: Detecting anomalous cells with M2ASDA package](./Anomaly.ipynb)
- [Tutorial 2: Running M2ASDA to detect anomalous cells in terminal](./Anomaly_T.ipynb)
